src/elfpy/markets.py

```diff
@@ -471,24 +428,35 @@ def _open_long(
         pricing_model: PricingModel,
         agent_action: MarketAction,
         token_out: TokenType,
-        stretched_time_remaining: float,
+        time_remaining: StretchedTime,
     ) -> tuple[MarketDeltas, Wallet]:
         """
         take trade spec & turn it into trade details
         compute wallet update spec with specific details
             will be conditional on the pricing model
         """
-        if agent_action.trade_amount <= self.bond_reserves:
+        # TODO: Why are we clamping elsewhere but we don't apply the trade at
```

Comment:
I agree, I think we shouldn't do either. I think it is too opaque to the end user. The agent should check this before making the trade. Also I think a realistic response would be that the market rejects the trade and informs the agent (& user). We could set a market variable  that allows "illegal" trades, for testing or whatever.

Reply:
I also like the approach of all of our components being strict about their inputs. This will serve us well if we ever try to run the simulation framework on our smart contracts, which will certainly be strict about the inputs that they accept.